### ISSUE TITLE
Upstream GGML patches and deprecate patched ggml.

### ISSUE THREAD

**Issue by mmwillet** [mmwillet/TTS.cpp#66]
Initial discussion for this came up [here](https://github.com/mmwillet/TTS.cpp/pull/65) and the current proposal, to upstream necessary operations directly to GGML or otherwise support them with custom operations and no longer require the patched version of ggml, represents the best long term outcome for TTS.cpp. 

Doing so will allow TTS.cpp to reap the benefits of recent GGML updates and make compatibility with other GGML reliant technologies more seamless.

Our goal would be, as @tyrauber outlined, to upstream the following support to GGML:

> Enhanced Conv Transpose 1D
> Remove the p0 == 0 constraint and add support for non-zero padding. Maintain backward compatibility (existing calls with p0=0 work unchanged). This is the most critical blocker and affects multiple model types. The lack of full PyTorch/TensorFlow parity in GGML's convolution operations seems like a significant gap that would benefit the broader ecosystem.

This should also include support for non-zero output padding (which is straightforward to add).

> Convolution Grouping Support
> Add grouping parameter to convolution operations. Required by Kokoro and common in modern CNN architectures. Extend existing conv operations with optional grouping parameter.

Our requirement here would only be that the grouping parameter can be applied to convolutional transposition.
 
> Enhanced Upscaling Operation
> Extend ggml_upscale with multiple interpolation modes. Better PyTorch/TensorFlow parity, removes need for custom ggml_upscale_linear. Maybe nearest (existing), linear (1D sequences), bilinear, bicubic, trilinear, area, nearest-exact.

Additional modifications such as `ggml_reciprocal`, STFT, and iSTFT, would also be desirable but could initially be supported via custom map operations.

Please see the [discussion linked above](https://github.com/mmwillet/TTS.cpp/pull/65) for posterity.

**Comment by mmwillet** [comment]
Added documentation on the ggml patch requirement [here](https://github.com/mmwillet/TTS.cpp/commit/3a81b94e33d34ae9d3263cf4edd910d44493dfa0).

**Comment by mmwillet** [comment]
@danielzgtg tested out the stride hack that you mentioned and it doesn't work because `ggml_new_tensor_impl` [raises an assert error when the produce of the tensor shape is 0.](https://github.com/mmwillet/ggml/blob/6e0934d350a7fb924606de33cd37279190c58e67/src/ggml.c#L1544)

**Comment by danielzgtg** [comment]
> stride hack [...] doesn't work

```diff
--- a/src/util.cpp
+++ b/src/util.cpp
@@ -79,7 +79,12 @@ float round_to_float(double v) {
 // Snake1d is a common tunable activation function used in the DAC model.
 struct ggml_tensor * snake_1d(ggml_context * ctx, struct ggml_tensor * alpha, struct ggml_tensor * a) {
     assert(a->ne[2] == 1 && a->ne[3] == 1);
-    return ggml_add(ctx, a, ggml_mul(ctx, ggml_sqr(ctx, ggml_sin(ctx, ggml_mul(ctx, a, alpha))), ggml_reciprocal(ctx, alpha)));
+    static constexpr float one = 1;
+    ggml_tensor * ones = ggml_new_tensor_2d(ctx, GGML_TYPE_F32, 1, alpha->ne[1]);
+    ones->nb[1] = 0; // broadcast
+    ones->data = const_cast<float *>(&one);
+
+    return ggml_add(ctx, a, ggml_mul(ctx, ggml_sqr(ctx, ggml_sin(ctx, ggml_mul(ctx, a, alpha))), ggml_div(ctx, ones, alpha)));
 }
 
 bool has_suffix(std::string value, std::string suffix) {
```

Output sounds fine and doesn't fail any asserts.

**Comment by danielzgtg** [comment]
I have a branch with the snake1d kernel fused: optimize/snake1d-fused-failed (4cdb85d0f0b2556b16440a1e7d3e078f095d7fc3). The optimization was statistically insignificant, and I must have already squeezed out all the potential gains with my ffast-math ggml_sin PR. The bottleneck is `convs1_weights` `build_kokoro_generator_res_block`. The math wasn't simplifiable, and a dilation of 5+ rules out SIMD. This instead creates demand for GPU support, which is good at matmuls.

`ggml_mul(ctx, other, ggml_reciprocal(ctx, alpha))` can be simplified to `ggml_div(ctx, other, alpha)` in standard ggml. However, the benefit of reciprocals is that multiplication is faster than division, and there is a specific RCPSS x86 instruction for theme, as well as reciprocals being able to be broadcasted from `[K]` to `[IC, K]`.

**Comment by henk717** [comment]
Had a chat with someone who contributes to llamacpp upstream, and he pointed me to this : https://github.com/ggml-org/llama.cpp/blob/master/docs/ops.md
That is the ops tracker that informs GPU backend developers if their own GPU backend is missing acceleration.
So if you focus on upstreaming your ops to the wider GGML project GPU support should be easier to achieve afterwards as its then not your project having to do all this work by yourself, your ops will appear as missing GPU backend ops by upstreaming them.
